neps/nep-519-yield-execution.md

```diff
@@ -0,0 +1,164 @@
+
+Examples include when a smart contract (`S`) provides MPC signing capabilities parties external to the NEAR protocol are computing the signature.  The rough steps are:
+1. Signer contract provides a function `fn sign_payload(Payload, ...)`.
+2. When called, the contract updates some contract state which is being monitored by external indexers to indicate that a new signing request has been received.  It also defers replying to the caller.
+3. The indexers observe the new signing request, they compute a signature and call another function `fn signature_available(Signature, ...)` on the signer contract.
+4. The signer contract validates the signature and if validate, replies to the original caller.
+
+Today, the NEAR protocol has no sensible way to defer replying to the caller in step 2 above.  This proposal proposes adding two following new host functions to the NEAR protocol:
+
+- `yield`: this can be called by a contract to indicate to the protocol that it is not ready yet to reply to its caller.
+- `resume`: a contract can use this mechanism to indicate to a protocol that it is now ready to reply to a caller that it had deferred earlier.
+
+If these two host functions were available, then `yield` would be used in step 2 above and `resume` would be used in step 4 above.
+
+## Specification
+
+The proposal is to add the following host functions to the NEAR protocol:
+
+
+```rust
+/// Instructs the protocol that the smart contract is not ready yet to respond
+/// to its caller yet.  The smart contract promises to call `yield_resume()`
+/// within `yield_num_blocks` blocks.  When `yield_resume()` is called, the
+/// protocol will call the method on the smart contract that is identified by
+/// `method_name_len` and `method_name_ptr` and this method will be expected to
+/// either respond to the caller or create another promise.
+///
+/// If the contract fails to call `yield_resume()` within `yield_num_blocks`,
+/// then the protocol will call the method on the smart contract that is
+/// identified by `method_name_len` and `method_name_ptr` with a timeout error.
+///
+/// `gas_for_resumed_method` is the prepayment of Gas that will be used to
+/// execute the method identified by `method_name_len` and `method_name_ptr`.
+///
+/// `gas_weight`: as specified in
+/// [this](https://github.com/near/NEPs/blob/master/neps/nep-0264.md) NEP, this
+/// improves the devX by allowing the developer to specify how to divide up the
+/// remaining gas.
+///
+/// Return value: u64: Similar to the
+/// [promise_create](https://github.com/near/nearcore/blob/a908de36ab6f75eb130447a5788007e26d05f93e/runtime/near-vm-runner/src/logic/logic.rs#L1281)
+/// host function, this function also create a promise and returns an index to
+/// the promise.
+pub fn yield_create(
+    method_name_len: u64,
+    method_name_ptr: u64,
```

Comment:
Why have the method name specified here rather than in `yield_resume`? Since we already have to validate the input of `yield_resume` this change would appear to make this strictly more powerful.

Happy to keep it this way if it's easier to implement though.

Reply:
I'm happy to make this change.  I don't think it will make the implementation trickier.  I don't even remember why I chose it this way now.

Reply:
I had written it this way because it made sense to me for the prepaid gas amount to be specified in the same place as the method which is being prepaid for. If we separate them the smart contract author will have to think about whether the gas prepaid in one place (`yield_create`) will be sufficient for the method call specified in a different place (`yield_resume`). 

That said, overall I do like the added flexibility and am happy with the change.

Reply:
This might not work for a different reason; we need to know the signature of the callback method to support composability with other promises. For example if we wish to compile something like `promise_create(...).and_then(...)`, we should know the type returned by the promise callback.

Reply:
Believe this is resolved now.